aeneas/src/ir/SsaNormalizer.v3

```diff
@@ -346,6 +346,179 @@ class SsaRaNormalizer extends SsaRebuilder {
+				}
+				return mapN(app, newArgs);
+			}
+			RefStructIn(offset) => {
+				var an = ArrayNorm.!(normType(V3.arrayByteType));
+				var newArgs = genRefs(app.inputs);
+				var array = newArgs[0], pos = newArgs[an.size];
+				newArgs[an.size] = curBlock.opIntAdd(newGraph.intConst(offset), pos);
+				return mapN(app, newArgs);
+			}
+			RefStructGetField(offset) => {
+				var an = ArrayNorm.!(normType(V3.arrayByteType));
+				var ortype = normReturnType(op).oldType, nrtype = normReturnType(op).newType;
+				var newArgs = genRefs(app.inputs);
+				var array = newArgs[0], pos = newArgs[an.size];
+				var result = SsaInstr.!(newGraph.intConst(0));
+				match (ortype) {
+					x: IntType => {
+						var width = x.packedByteSize();
+						result = genGetBytesInstr(app, array, pos,
+							offset, width, true, x, result);
+						return map1(app, result);
+					}
+					x: EnumType => {
+						var width = x.packedByteSize();
+						var enumType = x.enumDecl.tagType;
+						result = genGetBytesInstr(app, array, pos,
+							offset, width, true, Int.TYPE, result);
+						var enumTagLen = newGraph.intConst(x.enumDecl.cases.length);
+						var inBound = curBlock.opIntULt(IntType.!(result.getType()),
+							Int.TYPE, result, enumTagLen);
+						var split = SsaBlockSplit.new(context, curBlock);
+						curBlock = split.addIf(inBound);
+						curBlock = split.addElse();
+						curBlock = split.finish();
+						result = split.addPhi(enumType, [result, newGraph.longConst(0)]);
+						var intViewOp = V3Op.newIntViewI(result.getType(), nrtype);
+						result = curBlock.addApply(app.source, intViewOp, [result]);
+						return map1(app, result);
+					}
+					x: FloatType => {
+						var width = x.byteSize();
+						result = genGetBytesInstr(app, array, pos,
+							offset, width, true,  Int.getType(false, x.total_width), result);
+						var floatViewOp: Operator;
+						if (x.is64) {
+							floatViewOp = V3Op.newFloat64ViewI(result.getType());
+						}
+						else {
+							var intViewOp = V3Op.newIntViewI(result.getType(), Int.TYPE);
+							result = curBlock.addApply(app.source, intViewOp, [result]);
+							floatViewOp = V3Op.newFloat32ViewI(result.getType());
+						}
+						result = curBlock.addApply(app.source, floatViewOp, [result]);
+						return map1(app, result);
+					}
+					_ => ;
+				}
+			}
+			RefStructSetField(offset) => {
+				var newArgs = genRefs(args);
+				var array = newArgs[0], pos = newArgs[1], val = newArgs[2];
+				match(normTypeArg(op, 1).newType) {
+					x: IntType => {
+						var width = x.byteSize();
+						genSetBytesInstr(app, val, width, array, pos, offset, true, Int.TYPE);
+					}
+					x: EnumType => {
+						var width = x.byteSize();
+						val = curBlock.opIntViewI0(newArgs[2].getType(), Int.TYPE, val);
+						genSetBytesInstr(app, val, width, array, pos, offset, true, Int.TYPE);
+					}
+					x: FloatType => {
+						var width = x.byteSize();
+						val = curBlock.opIntView(val.getType(), Long.TYPE, val);
```

Comment:
Here I think we should use `Int.getType(false, x.total_width)` instead of `Long.TYPE`. Also I think you can just use `x` instead of `val.getType()`.